lib/Twig.php

```diff
@@ -22,12 +22,173 @@ class Twig {
 	public static function init() {
 		$self = new self();
 
-        add_action( 'timber/twig/filters', array( $self, 'add_timber_filters' ) );
-		add_action( 'timber/twig/functions', array( $self, 'add_timber_functions' ) );
-		add_action( 'timber/twig/escapers', array( $self, 'add_timber_escapers' ) );
+        add_action( 'timber/twig/filters', [ $self, 'add_timber_filters' ] );
+		add_action( 'timber/twig/functions', [ $self, 'add_timber_functions' ] );
+		add_action( 'timber/twig/escapers', [ $self, 'add_timber_escapers' ] );
 
         add_filter( 'timber/loader/twig', [ $self, 'set_defaults' ] );
-  }
+	}
+
+	public function get_timber_functions() {
+
+		$postFactory = new PostFactory();
+
+		return apply_filters( 'timber/twig/default_functions', [
+			[
+				'name' => 'action',
```

Comment:
Did you choose a non-associative array on purpose? What if we would choose the function name as the key? For example:

```php
[
    'action' => [
        'callable' => function( $action_name, ...$args ) {
		    do_action_ref_array( $action_name, $args );
		}
    ]
]
```

This would also make it easier to remove or replace a function using the `timber/twig/default_functions` filter:

```php
add_filter( 'timber/twig/default_functions', function( $functions ) {
    // Remove.
    unset( $functions['action'] );

    // Or replace.
    $functions['action'] = [
        'callable' => 'other',
    ];

    return $functions;
} );
```

Reply:
I did. I first wrote the function name as key and the callable as value. While checking Twig's docs, I noticed `\Twig\TwigFunction` could have a third `$options` parameter that could be used to deprecate functions and filters and would avoid repeating deprecation and thought that it would be more consistent to have an array that reflects `TwigFunction` arguments.

Remove or adding functions could be quite easy too:
```php
// Remove.
$functions = array_filter($functions, function($fn) {
    return !in_array('action', $fn['name']);
});
// Add
$functions[] = [
    'name' => 'wpautop',
    'callable' => 'wpautop',
];
```

But I didn't consider replacement which would require a bit more code 😅

Maybe you're right, function name as key would be more convenient.

Reply:
Yeah, it’s great that we can pass more parameters.

I just think that finding and filtering out entries that have their key in the array itself is not as straightforward as when we have keys. Yes, there are so many ways you can filter that array, and I think they are all a little too complicated:

- Using `array_filter()`, like you mentioned
- A foreach loop
- Using `array_search()`
- Using `wp_list_filter()`

And of course – I didn’t realize this either – replacing an item is much much harder. :)